Failed to pick up right Node.js version in sub folder

### How are you running Renovate?

Mend Renovate hosted app on github.com

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

This is the pull request having issue: https://github.com/Hongbo-Miao/hongbomiao.com/pull/5997

I have a mono repo. The sub folder `/grafana/hm-panel-plugin` expects using Node.js 16.

However, when Renovate tries to update package at `/grafana/hm-panel-plugin/package.json`,
I got [error](https://github.com/Hongbo-Miao/hongbomiao.com/pull/5997#issuecomment-1312449243) from Reonvate:

```sh
npm ERR! code EBADENGINE
npm ERR! engine Unsupported engine
npm ERR! engine Not compatible with your version of node/npm: hm-panel-plugin@1.0.0
npm ERR! notsup Not compatible with your version of node/npm: hm-panel-plugin@1.0.0
npm ERR! notsup Required: {"node":"16.x","npm":"8.x"}
npm ERR! notsup Actual:   {"npm":"8.19.3","node":"v18.12.1"}

npm ERR! A complete log of this run can be found in:
npm ERR!     /tmp/renovate-cache/others/npm/_logs/2022-11-12T18_31_35_449Z-debug-0.log
```

I think Renovate probably used the Node.js version 18 in the root [/.nvmrc](https://github.com/Hongbo-Miao/hongbomiao.com/blob/main/.nvmrc) instead of Node.js version 16 at [/grafana/hm-panel-plugin/.nvmrc](https://github.com/Hongbo-Miao/hongbomiao.com/blob/main/grafana/hm-panel-plugin/.nvmrc)

Let me know if need any other info, thanks! 😃

### Relevant debug logs

<details><summary>Logs</summary>

Part of log at https://app.renovatebot.com/dashboard#github/Hongbo-Miao/hongbomiao.com/886333983

```sh
lock file error(branch="renovate/tanstack-query-monorepo")
{
  "err": {
    "name": "ExecError",
    "cmd": "/bin/sh -c docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/mnt/renovate/gh/Hongbo-Miao/hongbomiao.com\":\"/mnt/renovate/gh/Hongbo-Miao/hongbomiao.com\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -v \"/tmp/containerbase\":\"/tmp/containerbase\" -e NPM_CONFIG_CACHE -e npm_config_store -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/mnt/renovate/gh/Hongbo-Miao/hongbomiao.com/grafana/hm-panel-plugin\" docker.io/renovate/sidecar bash -l -c \"install-tool node 18.12.1 && install-tool npm 8.19.3 && hash -d npm 2>/dev/null || true && npm install --package-lock-only --no-audit --ignore-scripts\"",
    "stderr": "npm ERR! code EBADENGINE\nnpm ERR! engine Unsupported engine\nnpm ERR! engine Not compatible with your version of node/npm: hm-panel-plugin@1.0.0\nnpm ERR! notsup Not compatible with your version of node/npm: hm-panel-plugin@1.0.0\nnpm ERR! notsup Required: {\"node\":\"16.x\",\"npm\":\"8.x\"}\nnpm ERR! notsup Actual:   {\"npm\":\"8.19.3\",\"node\":\"v18.12.1\"}\n\nnpm ERR! A complete log of this run can be found in:\nnpm ERR!     /tmp/renovate-cache/others/npm/_logs/2022-11-12T08_26_30_850Z-debug-0.log\n",
    "stdout": "installing v2 tool node v18.12.1\nlinking tool node v18.12.1\nnode: v18.12.1 /usr/local/bin/node\nnpm: 8.19.2  /usr/local/bin/npm\nInstalled v2 /usr/local/buildpack/tools/v2/node.sh in 7 seconds\nskip cleanup, not a docker build: 96b4d4cb4ab1\ninstalling v2 tool npm v8.19.3\n\nadded 1 package in 6s\nlinking tool npm v8.19.3\n8.19.3\nInstalled v2 /usr/local/buildpack/tools/v2/npm.sh in 8 seconds\nskip cleanup, not a docker build: 96b4d4cb4ab1\n",
    "options": {
      "cwd": "/mnt/renovate/gh/Hongbo-Miao/hongbomiao.com/grafana/hm-panel-plugin",
      "encoding": "utf-8",
      "env": {
        "NPM_CONFIG_CACHE": "/tmp/renovate-cache/others/npm",
        "npm_config_store": "/tmp/renovate-cache/others/pnpm",
        "HOME": "/home/user",
        "PATH": "/home/user/.local/bin:/home/user/bin:/opt/buildpack/tools/python/3.9.3/bin:/home/user/.npm-global/bin:/home/ubuntu/renovateapp/node_modules/.bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
        "LC_ALL": "C.UTF-8",
        "LANG": "C.UTF-8",
        "BUILDPACK_CACHE_DIR": "/tmp/containerbase",
        "CONTAINERBASE_CACHE_DIR": "/tmp/containerbase"
      },
      "maxBuffer": 10485760,
      "timeout": 900000
    },
    "exitCode": 1,
    "message": "Command failed: docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/mnt/renovate/gh/Hongbo-Miao/hongbomiao.com\":\"/mnt/renovate/gh/Hongbo-Miao/hongbomiao.com\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -v \"/tmp/containerbase\":\"/tmp/containerbase\" -e NPM_CONFIG_CACHE -e npm_config_store -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/mnt/renovate/gh/Hongbo-Miao/hongbomiao.com/grafana/hm-panel-plugin\" docker.io/renovate/sidecar bash -l -c \"install-tool node 18.12.1 && install-tool npm 8.19.3 && hash -d npm 2>/dev/null || true && npm install --package-lock-only --no-audit --ignore-scripts\"\nnpm ERR! code EBADENGINE\nnpm ERR! engine Unsupported engine\nnpm ERR! engine Not compatible with your version of node/npm: hm-panel-plugin@1.0.0\nnpm ERR! notsup Not compatible with your version of node/npm: hm-panel-plugin@1.0.0\nnpm ERR! notsup Required: {\"node\":\"16.x\",\"npm\":\"8.x\"}\nnpm ERR! notsup Actual:   {\"npm\":\"8.19.3\",\"node\":\"v18.12.1\"}\n\nnpm ERR! A complete log of this run can be found in:\nnpm ERR!     /tmp/renovate-cache/others/npm/_logs/2022-11-12T08_26_30_850Z-debug-0.log\n",
    "stack": "ExecError: Command failed: docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/mnt/renovate/gh/Hongbo-Miao/hongbomiao.com\":\"/mnt/renovate/gh/Hongbo-Miao/hongbomiao.com\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -v \"/tmp/containerbase\":\"/tmp/containerbase\" -e NPM_CONFIG_CACHE -e npm_config_store -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/mnt/renovate/gh/Hongbo-Miao/hongbomiao.com/grafana/hm-panel-plugin\" docker.io/renovate/sidecar bash -l -c \"install-tool node 18.12.1 && install-tool npm 8.19.3 && hash -d npm 2>/dev/null || true && npm install --package-lock-only --no-audit --ignore-scripts\"\nnpm ERR! code EBADENGINE\nnpm ERR! engine Unsupported engine\nnpm ERR! engine Not compatible with your version of node/npm: hm-panel-plugin@1.0.0\nnpm ERR! notsup Not compatible with your version of node/npm: hm-panel-plugin@1.0.0\nnpm ERR! notsup Required: {\"node\":\"16.x\",\"npm\":\"8.x\"}\nnpm ERR! notsup Actual:   {\"npm\":\"8.19.3\",\"node\":\"v18.12.1\"}\n\nnpm ERR! A complete log of this run can be found in:\nnpm ERR!     /tmp/renovate-cache/others/npm/_logs/2022-11-12T08_26_30_850Z-debug-0.log\n\n    at ChildProcess.<anonymous> (/home/ubuntu/renovateapp/node_modules/renovate/dist/util/exec/common.js:87:24)\n    at ChildProcess.emit (node:events:525:35)\n    at ChildProcess.emit (node:domain:489:12)\n    at Process.ChildProcess._handle.onexit (node:internal/child_process:291:12)"
  },
  "type": "npm"
}
```

</details>


### Have you created a minimal reproduction repository?

No reproduction, but I have linked to a public repo where it occurs

Here is the log _for the branch_:

```
DEBUG: syncBranchState()(branch="renovate/tanstack-query-monorepo")
DEBUG: syncBranchState(): Branch cache not found, creating minimal branchState(branch="renovate/tanstack-query-monorepo")
DEBUG: getBranchPr(renovate/tanstack-query-monorepo)(branch="renovate/tanstack-query-monorepo")
DEBUG: findPr(renovate/tanstack-query-monorepo, undefined, open)(branch="renovate/tanstack-query-monorepo")
DEBUG: findPr(renovate/tanstack-query-monorepo, undefined, closed)(branch="renovate/tanstack-query-monorepo")
DEBUG: branchExists=false(branch="renovate/tanstack-query-monorepo")
DEBUG: dependencyDashboardCheck=undefined(branch="renovate/tanstack-query-monorepo")
DEBUG: recreateClosed is false(branch="renovate/tanstack-query-monorepo")
DEBUG: findPr(renovate/tanstack-query-monorepo, fix(deps): update tanstack-query monorepo to v4.15.0, !open)(branch="renovate/tanstack-query-monorepo")
DEBUG: prAlreadyExisted=false(branch="renovate/tanstack-query-monorepo")
DEBUG: Checking schedule(at any time, null)(branch="renovate/tanstack-query-monorepo")
DEBUG: No schedule defined(branch="renovate/tanstack-query-monorepo")
DEBUG: Branch needs creating(branch="renovate/tanstack-query-monorepo")
DEBUG: Using reuseExistingBranch: false(branch="renovate/tanstack-query-monorepo")
DEBUG: Setting current branch to main(branch="renovate/tanstack-query-monorepo")
DEBUG: latest commit(branch="renovate/tanstack-query-monorepo")
{
  "branchName": "main",
  "latestCommitDate": "2022-11-11T23:42:09-08:00"
}
DEBUG: manager.getUpdatedPackageFiles() reuseExistingBranch=false(branch="renovate/tanstack-query-monorepo")
DEBUG: npm.updateDependency(): dependencies.@tanstack/react-query = 4.15.0(branch="renovate/tanstack-query-monorepo")
DEBUG: Updating @tanstack/react-query in web/package.json(branch="renovate/tanstack-query-monorepo")
DEBUG: npm.updateDependency(): dependencies.@tanstack/react-query = 4.15.0(branch="renovate/tanstack-query-monorepo")
DEBUG: Updating @tanstack/react-query in grafana/hm-panel-plugin/package.json(branch="renovate/tanstack-query-monorepo")
DEBUG: npm.updateDependency(): devDependencies.@tanstack/react-query-devtools = 4.15.0(branch="renovate/tanstack-query-monorepo")
DEBUG: Updating @tanstack/react-query-devtools in web/package.json(branch="renovate/tanstack-query-monorepo")
DEBUG: npm.updateDependency(): devDependencies.@tanstack/react-query-devtools = 4.15.0(branch="renovate/tanstack-query-monorepo")
DEBUG: Updating @tanstack/react-query-devtools in grafana/hm-panel-plugin/package.json(branch="renovate/tanstack-query-monorepo")
DEBUG: Updated 2 package files(branch="renovate/tanstack-query-monorepo")
DEBUG: Getting updated lock files(branch="renovate/tanstack-query-monorepo")
DEBUG: Writing package.json files(branch="renovate/tanstack-query-monorepo")
{
  "packageFiles": [
    "api-node/package.json",
    "cypress/package.json",
    "ethereum/package.json",
    "grafana/hm-panel-plugin/package.json",
    "mobile/package.json",
    "package.json",
    "web/package.json"
  ]
}
DEBUG: Writing api-node/package-lock.json(branch="renovate/tanstack-query-monorepo")
DEBUG: Writing cypress/package-lock.json(branch="renovate/tanstack-query-monorepo")
DEBUG: Writing ethereum/package-lock.json(branch="renovate/tanstack-query-monorepo")
DEBUG: Writing grafana/hm-panel-plugin/package-lock.json(branch="renovate/tanstack-query-monorepo")
DEBUG: Massaging npm lock file before writing to disk(branch="renovate/tanstack-query-monorepo")
DEBUG: Writing mobile/package-lock.json(branch="renovate/tanstack-query-monorepo")
DEBUG: Writing package-lock.json(branch="renovate/tanstack-query-monorepo")
DEBUG: Writing web/package-lock.json(branch="renovate/tanstack-query-monorepo")
DEBUG: Massaging npm lock file before writing to disk(branch="renovate/tanstack-query-monorepo")
DEBUG: Writing any updated package files(branch="renovate/tanstack-query-monorepo")
DEBUG: Writing web/package.json(branch="renovate/tanstack-query-monorepo")
DEBUG: Writing grafana/hm-panel-plugin/package.json(branch="renovate/tanstack-query-monorepo")
DEBUG: npmrc file found in repository(branch="renovate/tanstack-query-monorepo")
DEBUG: Generating package-lock.json for web(branch="renovate/tanstack-query-monorepo")
DEBUG: Spawning npm install to create web/package-lock.json(branch="renovate/tanstack-query-monorepo")
DEBUG: Updating lock file only(branch="renovate/tanstack-query-monorepo")
DEBUG: Using node constraint "18.12.1" from web/.nvmrc(branch="renovate/tanstack-query-monorepo")
DEBUG: Setting CONTAINERBASE_CACHE_DIR to /tmp/containerbase(branch="renovate/tanstack-query-monorepo")
DEBUG: Using docker to execute image: sidecar(branch="renovate/tanstack-query-monorepo")
DEBUG: Resolved stable matching version(branch="renovate/tanstack-query-monorepo")
{
  "toolName": "npm",
  "constraint": "8.x",
  "resolvedVersion": "8.19.3"
}
DEBUG: containerbaseDir is separate from cacheDir(branch="renovate/tanstack-query-monorepo")
DEBUG: No tag or tagConstraint specified for image: docker.io/renovate/sidecar(branch="renovate/tanstack-query-monorepo")
DEBUG: Docker image is already prefetched: docker.io/renovate/sidecar@sha256:5103d041fbc1e3deb7fd201ad0ed6ade56dc5c7f44c1613b094e58c022167fe8(branch="renovate/tanstack-query-monorepo")
DEBUG: exec completed(branch="renovate/tanstack-query-monorepo")
{
  "cmd": "docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/mnt/renovate/gh/Hongbo-Miao/hongbomiao.com\":\"/mnt/renovate/gh/Hongbo-Miao/hongbomiao.com\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -v \"/tmp/containerbase\":\"/tmp/containerbase\" -e NPM_CONFIG_CACHE -e npm_config_store -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/mnt/renovate/gh/Hongbo-Miao/hongbomiao.com/web\" docker.io/renovate/sidecar bash -l -c \"install-tool node 18.12.1 && install-tool npm 8.19.3 && hash -d npm 2>/dev/null || true && npm install --package-lock-only --no-audit --ignore-scripts\"",
  "durationMs": 32939,
  "stdout": "installing v2 tool node v18.12.1\nlinking tool node v18.12.1\nnode: v18.12.1 /usr/local/bin/node\nnpm: 8.19.2  /usr/local/bin/npm\nInstalled v2 /usr/local/buildpack/tools/v2/node.sh in 7 seconds\nskip cleanup, not a docker build: fa57930b8e63\ninstalling v2 tool npm v8.19.3\n\nadded 1 package in 6s\nlinking tool npm v8.19.3\n8.19.3\nInstalled v2 /usr/local/buildpack/tools/v2/npm.sh in 8 seconds\nskip cleanup, not a docker build: fa57930b8e63\n\nup to date in 14s\n\n339 packages are looking for funding\n  run `npm fund` for details\n",
  "stderr": "npm notice \nnpm notice New major version of npm available! 8.19.3 -> 9.1.1\nnpm notice Changelog: <https://github.com/npm/cli/releases/tag/v9.1.1>\nnpm notice Run `npm install -g npm@9.1.1` to update!\nnpm notice \n"
}
DEBUG: web/package-lock.json needs updating(branch="renovate/tanstack-query-monorepo")
DEBUG: npmrc file found in repository(branch="renovate/tanstack-query-monorepo")
DEBUG: Generating package-lock.json for grafana/hm-panel-plugin(branch="renovate/tanstack-query-monorepo")
DEBUG: Spawning npm install to create grafana/hm-panel-plugin/package-lock.json(branch="renovate/tanstack-query-monorepo")
DEBUG: Updating lock file only(branch="renovate/tanstack-query-monorepo")
DEBUG: Using node constraint "18.12.1" from web/.nvmrc(branch="renovate/tanstack-query-monorepo")
DEBUG: Setting CONTAINERBASE_CACHE_DIR to /tmp/containerbase(branch="renovate/tanstack-query-monorepo")
DEBUG: Using docker to execute image: sidecar(branch="renovate/tanstack-query-monorepo")
DEBUG: Resolved stable matching version(branch="renovate/tanstack-query-monorepo")
{
  "toolName": "npm",
  "constraint": "8.x",
  "resolvedVersion": "8.19.3"
}
DEBUG: containerbaseDir is separate from cacheDir(branch="renovate/tanstack-query-monorepo")
DEBUG: No tag or tagConstraint specified for image: docker.io/renovate/sidecar(branch="renovate/tanstack-query-monorepo")
DEBUG: Docker image is already prefetched: docker.io/renovate/sidecar@sha256:5103d041fbc1e3deb7fd201ad0ed6ade56dc5c7f44c1613b094e58c022167fe8(branch="renovate/tanstack-query-monorepo")
DEBUG: rawExec err(branch="renovate/tanstack-query-monorepo")
{
  "err": {
    "name": "ExecError",
    "cmd": "/bin/sh -c docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/mnt/renovate/gh/Hongbo-Miao/hongbomiao.com\":\"/mnt/renovate/gh/Hongbo-Miao/hongbomiao.com\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -v \"/tmp/containerbase\":\"/tmp/containerbase\" -e NPM_CONFIG_CACHE -e npm_config_store -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/mnt/renovate/gh/Hongbo-Miao/hongbomiao.com/grafana/hm-panel-plugin\" docker.io/renovate/sidecar bash -l -c \"install-tool node 18.12.1 && install-tool npm 8.19.3 && hash -d npm 2>/dev/null || true && npm install --package-lock-only --no-audit --ignore-scripts\"",
    "stderr": "npm ERR! code EBADENGINE\nnpm ERR! engine Unsupported engine\nnpm ERR! engine Not compatible with your version of node/npm: hm-panel-plugin@1.0.0\nnpm ERR! notsup Not compatible with your version of node/npm: hm-panel-plugin@1.0.0\nnpm ERR! notsup Required: {\"node\":\"16.x\",\"npm\":\"8.x\"}\nnpm ERR! notsup Actual:   {\"npm\":\"8.19.3\",\"node\":\"v18.12.1\"}\n\nnpm ERR! A complete log of this run can be found in:\nnpm ERR!     /tmp/renovate-cache/others/npm/_logs/2022-11-12T08_26_30_850Z-debug-0.log\n",
    "stdout": "installing v2 tool node v18.12.1\nlinking tool node v18.12.1\nnode: v18.12.1 /usr/local/bin/node\nnpm: 8.19.2  /usr/local/bin/npm\nInstalled v2 /usr/local/buildpack/tools/v2/node.sh in 7 seconds\nskip cleanup, not a docker build: 96b4d4cb4ab1\ninstalling v2 tool npm v8.19.3\n\nadded 1 package in 6s\nlinking tool npm v8.19.3\n8.19.3\nInstalled v2 /usr/local/buildpack/tools/v2/npm.sh in 8 seconds\nskip cleanup, not a docker build: 96b4d4cb4ab1\n",
    "options": {
      "cwd": "/mnt/renovate/gh/Hongbo-Miao/hongbomiao.com/grafana/hm-panel-plugin",
      "encoding": "utf-8",
      "env": {
        "NPM_CONFIG_CACHE": "/tmp/renovate-cache/others/npm",
        "npm_config_store": "/tmp/renovate-cache/others/pnpm",
        "HOME": "/home/user",
        "PATH": "/home/user/.local/bin:/home/user/bin:/opt/buildpack/tools/python/3.9.3/bin:/home/user/.npm-global/bin:/home/ubuntu/renovateapp/node_modules/.bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
        "LC_ALL": "C.UTF-8",
        "LANG": "C.UTF-8",
        "BUILDPACK_CACHE_DIR": "/tmp/containerbase",
        "CONTAINERBASE_CACHE_DIR": "/tmp/containerbase"
      },
      "maxBuffer": 10485760,
      "timeout": 900000
    },
    "exitCode": 1,
    "message": "Command failed: docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/mnt/renovate/gh/Hongbo-Miao/hongbomiao.com\":\"/mnt/renovate/gh/Hongbo-Miao/hongbomiao.com\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -v \"/tmp/containerbase\":\"/tmp/containerbase\" -e NPM_CONFIG_CACHE -e npm_config_store -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/mnt/renovate/gh/Hongbo-Miao/hongbomiao.com/grafana/hm-panel-plugin\" docker.io/renovate/sidecar bash -l -c \"install-tool node 18.12.1 && install-tool npm 8.19.3 && hash -d npm 2>/dev/null || true && npm install --package-lock-only --no-audit --ignore-scripts\"\nnpm ERR! code EBADENGINE\nnpm ERR! engine Unsupported engine\nnpm ERR! engine Not compatible with your version of node/npm: hm-panel-plugin@1.0.0\nnpm ERR! notsup Not compatible with your version of node/npm: hm-panel-plugin@1.0.0\nnpm ERR! notsup Required: {\"node\":\"16.x\",\"npm\":\"8.x\"}\nnpm ERR! notsup Actual:   {\"npm\":\"8.19.3\",\"node\":\"v18.12.1\"}\n\nnpm ERR! A complete log of this run can be found in:\nnpm ERR!     /tmp/renovate-cache/others/npm/_logs/2022-11-12T08_26_30_850Z-debug-0.log\n",
    "stack": "ExecError: Command failed: docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/mnt/renovate/gh/Hongbo-Miao/hongbomiao.com\":\"/mnt/renovate/gh/Hongbo-Miao/hongbomiao.com\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -v \"/tmp/containerbase\":\"/tmp/containerbase\" -e NPM_CONFIG_CACHE -e npm_config_store -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/mnt/renovate/gh/Hongbo-Miao/hongbomiao.com/grafana/hm-panel-plugin\" docker.io/renovate/sidecar bash -l -c \"install-tool node 18.12.1 && install-tool npm 8.19.3 && hash -d npm 2>/dev/null || true && npm install --package-lock-only --no-audit --ignore-scripts\"\nnpm ERR! code EBADENGINE\nnpm ERR! engine Unsupported engine\nnpm ERR! engine Not compatible with your version of node/npm: hm-panel-plugin@1.0.0\nnpm ERR! notsup Not compatible with your version of node/npm: hm-panel-plugin@1.0.0\nnpm ERR! notsup Required: {\"node\":\"16.x\",\"npm\":\"8.x\"}\nnpm ERR! notsup Actual:   {\"npm\":\"8.19.3\",\"node\":\"v18.12.1\"}\n\nnpm ERR! A complete log of this run can be found in:\nnpm ERR!     /tmp/renovate-cache/others/npm/_logs/2022-11-12T08_26_30_850Z-debug-0.log\n\n    at ChildProcess.<anonymous> (/home/ubuntu/renovateapp/node_modules/renovate/dist/util/exec/common.js:87:24)\n    at ChildProcess.emit (node:events:525:35)\n    at ChildProcess.emit (node:domain:489:12)\n    at Process.ChildProcess._handle.onexit (node:internal/child_process:291:12)"
  }
}
```

In this case it's the `node` version which is wrong: `Required: {\"node\":\"16.x\",\"npm\":\"8.x\"}\nnpm ERR! notsup Actual:   {\"npm\":\"8.19.3\",\"node\":\"v18.12.1\"}`

It seems to say:

```
DEBUG: Using node constraint "18.12.1" from web/.nvmrc(branch="renovate/tanstack-query-monorepo")
```

Any idea why it's using that folder?

I'm not sure we currently support mixed versions in the same branch. To debug it we'd want a simplified reproduction repo (instructions to follow)

> would it probably be an option to use the .nvmrc located relative to the lockfile

I'm in a similar situation and such a change would help. My monorepo looks like this

```
packages
  packageA/
      package.json
  packageB/
      package.json
apps
  appA/
      package.json
package.json
pnpm-lock.yaml
.nvmrc
```

Only the root `.nvmrc` specifies the Node.js version to be used across the repo (eg. not latest but `16.17.1`) but Renovate is using the latest version instead it fails to find any constraints from the current package `package.json` file siblings.

https://github.com/renovatebot/renovate/blob/49b7b733cd1ff5b80e57394fb1a814b1e7805df0/lib/modules/manager/npm/post-update/node-version.ts#L45


  ```
  {"name":"renovate","repository":"<rep>","branch":"renovate/google.maps-3.x","msg":"Generating pnpm-lock.yaml for .",v":0}
  {"name":"renovate","repository":"<rep>","branch":"renovate/google.maps-3.x","msg":"Spawning pnpm install to create pnpm-lock.yaml",v":0}
  {"name":"renovate","repository":"<rep>","branch":"renovate/google.maps-3.x","msg":"No node constraint found - using latest",v":0}
  {"name":"renovate","repository":"<rep>","branch":"renovate/google.maps-3.x","msg":"Setting CONTAINERBASE_CACHE_DIR to /renovate-manager-data/renovate-manager/renovate/cache/containerbase",v":0}
  {"name":"renovate","repository":"<rep>","branch":"renovate/google.maps-3.x","msg":"Using docker to execute image: sidecar",v":0}
  {"name":"renovate","repository":"<rep>","branch":"renovate/google.maps-3.x","toolName":"node","constraint":null,"resolvedVersion":"v18.12.1","msg":"Resolved stable matching version",v":0}
  {"name":"renovate","repository":"<rep>","branch":"renovate/google.maps-3.x","msg":"containerbaseDir is inside cacheDir",v":0}
  {"name":"renovate","repository":"<rep>","branch":"renovate/google.maps-3.x","msg":"No tag or tagConstraint specified for image: docker.io/renovate/sidecar",v":0}
  {"name":"renovate","repository":"<rep>","branch":"renovate/google.maps-3.x","msg":"Fetching Docker image: docker.io/renovate/sidecar",v":0}
  {"name":"renovate","repository":"<rep>","branch":"renovate/google.maps-3.x","msg":"Finished fetching Docker image docker.io/renovate/sidecar@sha256:5103d041fbc1e3deb7fd201ad0ed6ade56dc5c7f44c1613b094e58c022167fe8",v":0}
  {"name":"renovate","repository":"<rep>","branch":"renovate/google.maps-3.x","err":{"name":"ExecError","cmd":"/bin/sh -c docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/renovate-manager-data/renovate-manager/renovate/base/repos/github/\":\"/renovate-manager-data/renovate-manager/renovate/base/repos/github/<rep>\" -v \"/renovate-manager-data/renovate-manager/renovate/cache\":\"/renovate-manager-data/renovate-manager/renovate/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/renovate-manager-data/renovate-manager/renovate/base/repos/github/<rep>\" docker.io/renovate/sidecar bash -l -c \"install-tool node v18.12.1 && install-tool pnpm 7.5.2 && pnpm install --recursive --lockfile-only --ignore-scripts --ignore-pnpmfile\"","stderr":"","stdout":"installing v2 tool node v18.12.1\nlinking tool node v18.12.1\nnode: v18.12.1 /usr/local/bin/node\nnpm: 8.19.2  /usr/local/bin/npm\nInstalled v2 /usr/local/buildpack/tools/v2/node.sh in 3 seconds\nskip cleanup, not a docker build: a839625bbc54\ninstalling v2 tool pnpm v7.5.2\n\nadded 1 package in 721ms\nlinking tool pnpm v7.5.2\n7.5.2\nInstalled v2 /usr/local/buildpack/tools/v2/pnpm.sh in 1 seconds\nskip cleanup, not a docker build: a839625bbc54\n ERR_PNPM_UNSUPPORTED_ENGINE  Unsupported environment (bad pnpm and/or Node.js version)\n\nYour Node version is incompatible with \"/renovate-manager-data/renovate-manager/renovate/base/repos/github/<rep>\".\n\nExpected version: >=16.17.1 <17\nGot: v18.12.1\n\nThis is happening because the package's manifest has an engines.node field specified.\nTo fix this issue, install the required Node version.\n","options":{"cwd":"/renovate-manager-data/renovate-manager/renovate/base/repos/github/<rep>","encoding":"utf-8","env":{"NPM_CONFIG_CACHE":"/renovate-manager-data/renovate-manager/renovate/cache/others/npm","npm_config_store":"/renovate-manager-data/renovate-manager/renovate/cache/others/pnpm","HOME":"/home/ubuntu","PATH":"/home/ubuntu/bin:/home/ubuntu/.npm-global/bin:/home/ubuntu/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin","LC_ALL":"C.UTF-8","LANG":"C.UTF-8","DOCKER_HOST":"127.0.0.1:2376","DOCKER_TLS_VERIFY":"true","DOCKER_CERT_PATH":"/docker/certs/client","BUILDPACK_CACHE_DIR":"/renovate-manager-data/renovate-manager/renovate/cache/containerbase","CONTAINERBASE_CACHE_DIR":"/renovate-manager-data/renovate-manager/renovate/cache/containerbase"},"maxBuffer":10485760,"timeout":600000},"exitCode":1,"message":"Command failed: docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/renovate-manager-data/renovate-manager/renovate/base/repos/github/<rep>\":\"/renovate-manager-data/renovate-manager/renovate/base/repos/github/<rep>\" -v \"/renovate-manager-data/renovate-manager/renovate/cache\":\"/renovate-manager-data/renovate-manager/renovate/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/renovate-manager-data/renovate-manager/renovate/base/repos/github/<rep>\" docker.io/renovate/sidecar bash -l -c \"install-tool node v18.12.1 && install-tool pnpm 7.5.2 && pnpm install --recursive --lockfile-only --ignore-scripts --ignore-pnpmfile\"\n","stack":"ExecError: Command failed: docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/renovate-manager-data/renovate-manager/renovate/base/repos/github/<rep>\":\"/renovate-manager-data/renovate-manager/renovate/base/repos/github/<rep>\" -v \"/renovate-manager-data/renovate-manager/renovate/cache\":\"/renovate-manager-data/renovate-manager/renovate/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/renovate-manager-data/renovate-manager/renovate/base/repos/github/<rep>\" docker.io/renovate/sidecar bash -l -c \"install-tool node v18.12.1 && install-tool pnpm 7.5.2 && pnpm install --recursive --lockfile-only --ignore-scripts --ignore-pnpmfile\"\n\n    at ChildProcess.<anonymous> (/usr/src/app/node_modules/renovate/lib/util/exec/common.ts:99:11)\n    at ChildProcess.emit (node:events:525:35)\n    at ChildProcess.emit (node:domain:489:12)\n    at Process.ChildProcess._handle.onexit (node:internal/child_process:293:12)"},"msg":"rawExec err","v":0}
  ```

--- a/lib/modules/manager/npm/post-update/lerna.ts
+++ b/lib/modules/manager/npm/post-update/lerna.ts
@@ -43,7 +43,7 @@ export async function generateLockFiles(
   }
   logger.debug(`Spawning lerna with ${lernaClient} to create lock files`);
   const toolConstraints: ToolConstraint[] = [
-    await getNodeToolConstraint(config, []),
+    await getNodeToolConstraint(config, [], lockFileDir),
   ];
   const cmd: string[] = [];
   let cmdOptions = '';

--- a/lib/modules/manager/npm/post-update/node-version.spec.ts
+++ b/lib/modules/manager/npm/post-update/node-version.spec.ts
@@ -17,34 +17,34 @@ describe('modules/manager/npm/post-update/node-version', () => {
     it('returns package.json range', async () => {
       fs.readLocalFile.mockResolvedValueOnce(null as never);
       fs.readLocalFile.mockResolvedValueOnce(null as never);
-      const res = await getNodeConstraint(config);
+      const res = await getNodeConstraint(config, '');
       expect(res).toBe('^12.16.0');
     });
 
     it('returns .node-version value', async () => {
       fs.readLocalFile.mockResolvedValueOnce(null as never);
       fs.readLocalFile.mockResolvedValueOnce('12.16.1\n');
-      const res = await getNodeConstraint(config);
+      const res = await getNodeConstraint(config, '');
       expect(res).toBe('12.16.1');
     });
 
     it('returns .nvmrc value', async () => {
       fs.readLocalFile.mockResolvedValueOnce('12.16.2\n');
-      const res = await getNodeConstraint(config);
+      const res = await getNodeConstraint(config, '');
       expect(res).toBe('12.16.2');
     });
 
     it('ignores unusable ranges in dotfiles', async () => {
       fs.readLocalFile.mockResolvedValueOnce('latest');
       fs.readLocalFile.mockResolvedValueOnce('lts');
-      const res = await getNodeConstraint(config);
+      const res = await getNodeConstraint(config, '');
       expect(res).toBe('^12.16.0');
     });
 
     it('returns no constraint', async () => {
       fs.readLocalFile.mockResolvedValueOnce(null as never);
       fs.readLocalFile.mockResolvedValueOnce(null as never);
-      const res = await getNodeConstraint({ ...config, constraints: null });
+      const res = await getNodeConstraint({ ...config, constraints: null }, '');
       expect(res).toBeNull();
     });
   });
@@ -64,9 +64,11 @@ describe('modules/manager/npm/post-update/node-version', () => {
   describe('getNodeToolConstraint()', () => {
     it('returns getNodeUpdate', async () => {
       expect(
-        await getNodeToolConstraint(config, [
-          { depName: 'node', newValue: '16.15.0' },
-        ])
+        await getNodeToolConstraint(
+          config,
+          [{ depName: 'node', newValue: '16.15.0' }],
+          ''
+        )
       ).toEqual({
         toolName: 'node',
         constraint: '16.15.0',
@@ -74,7 +76,7 @@ describe('modules/manager/npm/post-update/node-version', () => {
     });
 
     it('returns getNodeConstraint', async () => {
-      expect(await getNodeToolConstraint(config, [])).toEqual({
+      expect(await getNodeToolConstraint(config, [], '')).toEqual({
         toolName: 'node',
         constraint: '^12.16.0',
       });

--- a/lib/modules/manager/npm/post-update/node-version.ts
+++ b/lib/modules/manager/npm/post-update/node-version.ts
@@ -1,7 +1,8 @@
 import semver from 'semver';
+import upath from 'upath';
 import { logger } from '../../../../logger';
 import type { ToolConstraint } from '../../../../util/exec/types';
-import { getSiblingFileName, readLocalFile } from '../../../../util/fs';
+import { readLocalFile } from '../../../../util/fs';
 import { newlineRegex, regEx } from '../../../../util/regex';
 import type { PostUpdateConfig, Upgrade } from '../../types';
 
@@ -33,13 +34,13 @@ function getPackageJsonConstraint(
 }
 
 export async function getNodeConstraint(
-  config: Partial<PostUpdateConfig>
+  config: Partial<PostUpdateConfig>,
+  lockFileDir: string
 ): Promise<string | null> {
-  const { packageFile } = config;
   // TODO: fix types (#7154)
   const constraint =
-    (await getNodeFile(getSiblingFileName(packageFile!, '.nvmrc'))) ??
-    (await getNodeFile(getSiblingFileName(packageFile!, '.node-version'))) ??
+    (await getNodeFile(upath.join(lockFileDir, '.nvmrc'))) ??
+    (await getNodeFile(upath.join(lockFileDir, '.node-version'))) ??
     getPackageJsonConstraint(config);
   if (!constraint) {
     logger.debug('No node constraint found - using latest');
@@ -53,10 +54,11 @@ export function getNodeUpdate(upgrades: Upgrade[]): string | undefined {
 
 export async function getNodeToolConstraint(
   config: Partial<PostUpdateConfig>,
-  upgrades: Upgrade[]
+  upgrades: Upgrade[],
+  lockFileDir: string
 ): Promise<ToolConstraint> {
   const constraint =
-    getNodeUpdate(upgrades) ?? (await getNodeConstraint(config));
+    getNodeUpdate(upgrades) ?? (await getNodeConstraint(config, lockFileDir));
 
   return {
     toolName: 'node',

--- a/lib/modules/manager/npm/post-update/npm.ts
+++ b/lib/modules/manager/npm/post-update/npm.ts
@@ -63,7 +63,7 @@ export async function generateLockFile(
       cwdFile: lockFileName,
       extraEnv,
       toolConstraints: [
-        await getNodeToolConstraint(config, upgrades),
+        await getNodeToolConstraint(config, upgrades, lockFileDir),
         npmToolConstraint,
       ],
       docker: {

--- a/lib/modules/manager/npm/post-update/pnpm.ts
+++ b/lib/modules/manager/npm/post-update/pnpm.ts
@@ -46,7 +46,7 @@ export async function generateLockFile(
         image: 'sidecar',
       },
       toolConstraints: [
-        await getNodeToolConstraint(config, upgrades),
+        await getNodeToolConstraint(config, upgrades, lockFileDir),
         pnpmToolConstraint,
       ],
     };

--- a/lib/modules/manager/npm/post-update/yarn.ts
+++ b/lib/modules/manager/npm/post-update/yarn.ts
@@ -99,7 +99,7 @@ export async function generateLockFile(
   let lockFile: string | null = null;
   try {
     const toolConstraints: ToolConstraint[] = [
-      await getNodeToolConstraint(config, upgrades),
+      await getNodeToolConstraint(config, upgrades, lockFileDir),
     ];
     const yarnUpdate = upgrades.find(isYarnUpdate);
     const yarnCompatibility = yarnUpdate

